packages/fx-core/src/core/middleware/projectMigrator.ts

```diff
@@ -0,0 +1,427 @@
+  return { hasFrontend, hasBackend, hasBot, hasProvision };
+}
+
+async function getMultiEnvFolders(projectPath: string): Promise<any> {
+  const fx = path.join(projectPath, `.${ConfigFolderName}`);
+  const fxConfig = path.join(fx, InputConfigsFolderName);
+  const templateAppPackage = path.join(projectPath, "templates", AppPackageFolderName);
+  const fxPublishProfile = path.join(fx, PublishProfilesFolderName);
+  await fs.ensureDir(fxConfig);
+  await fs.ensureDir(templateAppPackage);
+  return { fx, fxConfig, templateAppPackage, fxPublishProfile };
+}
+
+async function removeOldProjectFiles(projectPath: string): Promise<void> {
+  const fx = path.join(projectPath, `.${ConfigFolderName}`);
+  await fs.remove(path.join(fx, "env.default.json"));
+  await fs.remove(path.join(fx, "default.userdata"));
+  await fs.remove(path.join(fx, "settings.json"));
+  await fs.remove(path.join(fx, "local.env"));
+  await fs.remove(path.join(projectPath, AppPackageFolderName));
+  await fs.remove(path.join(fx, "new.env.default.json"));
+  // version <= 2.4.1, rmove .fx/appPackage.
+  await fs.remove(path.join(fx, AppPackageFolderName));
+}
+
+async function ensureProgrammingLanguage(
+  projectSettingPath: string,
+  envDefaultPath: string
+): Promise<void> {
+  const settings: ProjectSettings = await readJson(projectSettingPath);
+  if (!settings.programmingLanguage || !settings.defaultFunctionName) {
+    const envDefault = await readJson(envDefaultPath);
+    settings.programmingLanguage = envDefault[PluginNames.SOLUTION][programmingLanguage];
+    settings.defaultFunctionName = envDefault[PluginNames.FUNC][defaultFunctionName];
+    await fs.writeFile(projectSettingPath, JSON.stringify(settings, null, 4), {
+      encoding: "UTF-8",
+    });
+  }
+}
+async function ensureActiveEnv(projectSettingPath: string): Promise<void> {
+  const settings: ProjectSettings = await readJson(projectSettingPath);
+  if (!settings.activeEnvironment) {
+    settings.activeEnvironment = "dev";
+    await fs.writeFile(projectSettingPath, JSON.stringify(settings, null, 4), {
+      encoding: "UTF-8",
+    });
+  }
+}
+
+async function cleanup(projectPath: string): Promise<void> {
+  const { _, fxConfig, templateAppPackage, fxPublishProfile } = await getMultiEnvFolders(
+    projectPath
+  );
+  await fs.remove(fxConfig);
+  await fs.remove(templateAppPackage);
+  await fs.remove(fxPublishProfile);
+  await fs.remove(path.join(templateAppPackage, ".."));
+  if (checkFileExist(path.join(fxConfig, "..", "new.env.default.json"))) {
+    await fs.remove(path.join(fxConfig, "..", "new.env.default.json"));
+  }
+}
+
+async function needMigrateToArmAndMultiEnv(ctx: CoreHookContext): Promise<boolean> {
+  if (!preCheckEnvEnabled()) {
+    return false;
+  }
+  const inputs = ctx.arguments[ctx.arguments.length - 1] as Inputs;
+  const fxExist = await fs.pathExists(path.join(inputs.projectPath as string, ".fx"));
+  if (!fxExist) {
```

Comment:
under what scenario a teamsfx project does not have ".fx" folder?

Reply:
I think this is a security check for the existence of a file
if user delete this .fx folder manually unexpected